### ISSUE TITLE
Consider adding Clock to JwtAccessTokenValidator

### ISSUE THREAD

**Issue by jzheaux** [jzheaux/spring-security-oauth2-resource-server#57]
Currently, this validator always compares against the system time (i.e. `Instant.now()`)

```java
if ( Instant.now().minus(maxClockSkew).isAfter(expiry) ) {
   ...
}
```
Though adding a configurable `Clock` would allow us to do:

```java
if ( Instant.now(this.clock).minus(maxClockSkew).isAfter(expiry) ) {
   ...
}
```
which can be helpful for testing, e.g. specifying a fixed clock or an offset clock.

**Comment by Clockwork-Muse** [comment]
Like I've encountered with a number of programmers, you forget that `Instant`... has no concept of timezone, and so isn't UTC nor not-UTC, it just represents a point on the timeline.  You can't actually deviate from UTC in that sense, because you're not actually dealing with it in the first place.

From a testing point of view, mostly what it would let you do is supply a fixed clock - one with a defined instant - which would allow predefined tokens to be used in testing (so no getting new signatures, among other things).

**Comment by Clockwork-Muse** [comment]
>So, while `Instant` in the abstract is independent of timezones, etc., as you say, `Instant.now()` is not. Right now, because the validator calls `Instant.now()` directly, it does assume UTC. Am I wrong?

You're wrong.  It's impossible to misuse this in terms of timezone for the validator.  All `Instant.now()` assumes is that you want to hit the system clock for the time.  That clock isn't "wall clock" or "displayed clock" (that is, local time), but rather the "absolute" clock; for most practical purposes the clock can be considered to reflect NTP.

Instants in the abstract are independent from UTC, and `Instant` in implementation is independent from UTC.  The implementation only contains seconds and nanos since the start of the standard Unix (or here, Java) epoch, and contains no reference to a timezone.  Calling `Instant.now(Clock.systemUTC())` and `Instant.now(Clock.systemDefaultZone())` will result in the same instant (obviously, assuming they're called at the exact same point in time).  Importantly, `Clock`s don't actually return "local" time - they return `Instant`s and a timezone; if a type is interested in local time, it derives it by applying the timezone offset (in effect at the given instant) to the instant, and stores that in a different domain type.

<hr />
"But wait!", you say, "toString() prints out an ISO UTC timestamp!  It must be tied to UTC!".

Sure it does - and while I don't know the real reason the behavior was chosen, I would choose to do so for at least the following reasons:

 1. It's tradition.  Instants have always been properly formatted as UTC.
 2. It's a relatively human-readable format, as most people will generally know their own offset.
 3. It roundtrips, and is unambiguous.  Which is important for de/serialization.

The important thing to consider is that while the Unix/Java epoch is formally defined as being `'1970-01-01 00:00:00.000000000Z'`.... that exact same instant in time maps to `'1969-12-31 16:00:00.000000000-08:00'`, and can be defined in those terms.

Pedantically, then, logging (among other things) isn't actually in UTC - you're representing a fixed point in time, an instant, and UTC isn't actually involved.  For _readability_ and human comprehension, when serializing it in text files it's often done to UTC for the reasons mentioned above, but some systems will just store the number of seconds.... like for OAuth tokens.

(I'm not going to touch the perils of scheduling future events, which should almost never be done as `Instant`s, because that's not completely relevant here)

**Comment by jzheaux** [comment]
Huh, okay. I hope you don't mind, but I did some extra digging to corroborate what you said as I think that `Instant.now(Clock.systemDefaultZone())` _reads_ like "get me the current instant, offset by the system default zone".

It turns out (as you already knew) that `SystemClock`, the implementation of `Clock` that is pertinent, simply does the following in its `instant()` method:

```java
@Override
public long millis() {
    return System.currentTimeMillis();
}

@Override
public Instant instant() {
    return Instant.ofEpochMilli(millis());
}
```
And doesn't take the provided `zone` into account at all. You are correct that `Clock.systemUTC().instant()` and `Clock.systemDefaultZone().instant()` both return the same instant insofar as they both simply call `System.currentTimeMillis()` (which you explained in your above comment).

Also, it says here in the [`Clock` java doc](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html):

> The primary purpose of this abstraction is to allow alternate clocks to be
> plugged in as and when required. Applications use an object to obtain the
> current time rather than a static method. This can simplify testing.
>
> Best practice for applications is to pass a {@code Clock} into any method
> that requires the current instant.

So, hey, thanks for teaching me something! This resolves my concern about accidental misuse.

**Comment by Clockwork-Muse** [comment]
>Huh, okay. I hope you don't mind, but I did some extra digging to corroborate what you said as I think that `Instant.now(Clock.systemDefaultZone())` reads like "get me the current instant, offset by the system default zone".

:shrug:
Eh, I don't mind.  I'm a bit more into the weeds on this area than most programmers (I've contributed to threeten-extra, and have looked into porting the library itself).  It makes me more pedantic about this stuff.

As for offsetting an instant by the timezone... I would say that that can't actually be done, by definition (ignoring API considerations).  This is about the one problem with making the display/format be a human-readable ISO UTC stamp, that people think timezone matters for this.  If timezone matters, you don't have an `Instant`, you have `ZonedDateTime`, where the zone is (rarely) UTC.

**Comment by jzheaux** [comment]
Fixed via 24b8ce0
